### DISCUSSION TITLE
Explain the meaning of the advice "Always use set to define a store" from the section "Flux inspired practice"

### DISCUSSION THREAD

**Discussion by budarin** [pmndrs/zustand#1863] [Q&A] [answered]
Does this mean that using no store actions does not comply with the principles of Flux?

Or is the meaning in something else?

**Comment by dai-shi** [comment]
As far as I know, I don't think we have a clear definition of Flux principles. That's why it's noted as "Flux inspired". One of the principles is one-way data flow, and store actions (that wrap `set`) are for it.

**Comment by budarin** [comment]
> "One of the principles is one-way data flow, and store actions (that wrap set) are for it."

this phrase causes misunderstanding and questions:
- do other methods not work the same way as set? 
- why?
- other methods are better not to use because they do not correspond to the spirit of Flux?

**Reply by dai-shi** [reply]
I'm confused with the questions. What would you suggest instead?
Because I'm not insisting any opinions about practices. I'm just trying to rephrase what community says.

**Reply by budarin** [reply]
I do not know what to offer - I just have a suggestion that raises questions to which I have no answers, so I decided to ask about it

**Reply by dai-shi** [reply]
That's fair. Let's have others and get more thoughts. @dbritto-dev @sewera @chrisk-7777

**Reply by budarin** [reply]
let's at least answer the questions that torment me in the context of this sentence: 

- is `set` somehow different from `store.setState`?
- does `store.setState` correspond to the spirit of Flux?

**Reply by dai-shi** [reply]
> let's at least answer the questions that torment me in the context of this sentence:

Those questions are clear to me. Thanks!

> * is `set` somehow different from `store.setState`?

They are identical.

> * does `store.setState` correspond to the spirit of Flux?

I don't believe so.

**Reply by budarin** [reply]
just reading this sentence, I have to think like this: never use no store actions - this does not correspond to the spirit of Flux and will not work the same way when using set

**Reply by budarin** [reply]
it seems to me that this sentence needs to be rephrased, As it is formulated now - it causes alarm and a lot of questions that I can do something wrong

**Reply by dai-shi** [reply]
Cool. I think you can make suggestions now? Feel free to open a PR.

**Reply by budarin** [reply]
I do not yet know the true meaning of this sentence  
how can I open an issue for it? :)

**Reply by dai-shi** [reply]
You don't need to open an issue. Please keep discussing here.

**Reply by budarin** [reply]
I made a typo - I meant PR

**Reply by dai-shi** [reply]
Do you mean where the docs are located?
Here: https://github.com/pmndrs/zustand/blob/main/docs/guides/flux-inspired-practice.md

**Reply by dbritto-dev** [reply]
@budarin inspired does not mean that should strict follow flux, just means that only takes the best parts from flux, that is why zustand is less opinionated than flux or any other flux-like alternatives

**Reply by budarin** [reply]
I have been using Flux and redux for a long time, but for me this phrase is absolutely not clear in the context of Zustand and only raises questions

**Reply by budarin** [reply]
> Do you mean where the docs are located? Here: https://github.com/pmndrs/zustand/blob/main/docs/guides/flux-inspired-practice.md

yes, that is it

**Reply by dai-shi** [reply]
> I have been using Flux and redux for a long time

Would love to hear your suggestion then!

**Reply by budarin** [reply]
I 'm not sure that I understand what you wanted to say with this phrase ...

**Reply by dai-shi** [reply]
hmm, what _I_ wanted to say was, using `store.setState` is against Flux principles. But, I'm not 100% sure if it's the true statement.

**Reply by budarin** [reply]
why? what's wrong with using `store.setState`?
Flux insists that the actions should be applied to the storage in order one after the other - there is no difference how the actions will be performed: using set or using `store.setState`

**Reply by dai-shi** [reply]
then, it can be my misunderstanding of Flux. I thought calling `store.setState()` isn't considered actions. There's nothing wrong from Zustand perspective. It's just trying to help Flux users migrating to Zustand. So, I'm open for your suggestions. (I'm fine to delete the entire page if it's too confusing. But, others may have other opinions.)

**Reply by budarin** [reply]
why calling store.setState() isn't considered actions?

I think that it is not worth deleting, but it is worth to formulate the idea more precisely so that it does not cause discrepancies and questions

**Reply by budarin** [reply]
in terms of Flux, action - it is simply an object containing data and metadata intended to be transferred to the store for subsequent modification of data in the store.
There is no such intermediate in Zustand, so that `set` and `store.setState` both perform an action

**Reply by dai-shi** [reply]
Okay, it's terminology. I thought functions like `increment()` are actions. If `set(...)` is an action, what's described would be different.

**Reply by chrisk-7777** [reply]
I think the closest equivalent term from Redux would be "dispatcher function"* or "dispatch function". 

The other thing to consider is **Redux** is just one implementation of **Flux** (albeit the most popular one!), with its own terminology. A library, like Zustand, can be Flux-inspired without being like Redux. 

Like @dbritto-dev mentioned, inspired doesn't mean to exactly follow. If we try to shoe horn terminology it might lead to more confusion. I guess I just want to be careful of equating terminology between the two libraries because they have different approaches. 

However, I do agree that a lot of people transitioning from another Flux-implementation (like Redux) might be unsure. 

I will take some time this week to update https://github.com/pmndrs/zustand/blob/main/docs/guides/flux-inspired-practice.md to clean up some of the terminology and highlight where Zustand and Redux/Flux are similar, but also where they differ. 

@budarin What would you say is the most obvious source of questions on that page? Especially after this discussion? Is it the use of `set`? Or the use "actions"? Something else?

**Reply by budarin** [reply]
The expression "Always use set to define storage" is confusing!
It seems to me that it would be important to write like this: to perform actions on data in the repository, use the store's `set` method and no store method - `setStore`

Actions in Flux terminology are data passed to `set` and `setState`
`Dispatch` is good analogy for `set` and `setState`

**Reply by chrisk-7777** [reply]
Thanks @budarin - Yes I think the word "define" is the wrong word in this case. `set` should be called to "update" the store, not "define" it. 

If we had something like:

> "Always use `set` (or `setState`) to perform updates on the store. `set` (and `setState`) ensures the described update is immutably performed on the store." 

Would that make it less confusing? 

Can you explain this line from your comment: "and no store method - `setStore`"? 
I'm not really understanding what you mean by "no store method", and when you say `setStore`, do you mean `setState`?

`set` and `setState` are essentially the same thing - there is some reading on the two approaches here: https://docs.pmnd.rs/zustand/guides/practice-with-no-store-actions

I just want to make sure I'm understanding so I can make the appropriate updates. 

Regarding "`Dispatch` is a good analogy" - I don't know if that aligns to `set`/`setState`. In Redux, dispatching an action is a "fire and forgot" process, a single action can be dispatched and 0 (zero), 1, or multiple reducers can "respond" to it. The dispatcher (as in, the caller of `dispatch()`) is unaware how a reducer responds (if at all). 

Whereas in Zustand, `set`/`setState` is more "direct" and performs the update there and then.

**Reply by budarin** [reply]
`setStore ` - is a typo, `setState` of course! )

>Whereas in Zustand, set/setState is more "direct" and performs the update there and then.

I agree

**Reply by chrisk-7777** [reply]
No worries about the typo @budarin . 

What are your thoughts on this line from my previous comment:

> "Always use set (or setState) to perform updates on the store. set (and setState) ensures the described update is immutably performed on the store."

Would that make the "define" line less confusing?

**Reply by budarin** [reply]
"define" is good

**Reply by chrisk-7777** [reply, marked as answer]
I have raised a PR here: https://github.com/pmndrs/zustand/pull/1874

@budarin can you have a look to see if it addresses your concerns?

**Reply by budarin** [reply]
Yes, quite enough

**Reply by chrisk-7777** [reply]
Fantastic, thanks for the feedback and continued conversation 👍

**Comment by chrisk-7777** [comment]
Thanks everyone for the solid conversation! PR #1874 has been merged, so I'll mark this discussion as resolved 🚀
